internal/socket.go

```diff
@@ -14,10 +14,142 @@ import (
+
+func isInAllowedGroup(ucred *unix.Ucred) (bool, error) {
+	userInfo, err := user.LookupId(fmt.Sprintf("%d", ucred.Uid))
+	if err != nil {
+		return false, fmt.Errorf("authenticate user, lookup user info: %s", err)
+	}
+	// user belongs to the allowed group?
+	groups, err := userInfo.GroupIds()
+	if err != nil {
+		return false, fmt.Errorf("authenticate user, check user groups: %s", err)
+	}
+
+	for _, groupId := range groups {
+		groupInfo, err := user.LookupGroupId(groupId)
+		if err != nil {
+			return false, fmt.Errorf("authenticate user, check user group: %s", err)
+		}
+		for _, allowGroupName := range allowedGroups {
+			if groupInfo.Name == allowGroupName {
+				return true, nil
+			}
+		}
+	}
+
+	return false, nil
+}
+
+// getUnixCreds returns info from unix socket connection about the process on the other end.
+func getUnixCreds(conn net.Conn, authenticator SocketAuthenticator) (*unix.Ucred, error) {
+	conn2 := extractConnection(conn)
+	if conn2 == nil {
+		return nil, fmt.Errorf("cannot extract connection of proper type, is netutil.LimitListener used?")
+	}
+
+	unixConn, ok := conn2.(*net.UnixConn)
+	if !ok {
+		return nil, fmt.Errorf("socket is not a unix socket")
+	}
+
+	rawConn, err := unixConn.SyscallConn()
+	if err != nil {
+		return nil, fmt.Errorf("getting raw connection: %w", err)
+	}
+
+	var ucred *unix.Ucred
+	var internalErr error
+	err = rawConn.Control(func(fd uintptr) {
+		ucred, internalErr = unix.GetsockoptUcred(int(fd), unix.SOL_SOCKET, unix.SO_PEERCRED)
+	})
+	if internalErr != nil {
+		return nil, fmt.Errorf("doing GetsockoptUcred: %w", internalErr)
+	}
+	if err != nil {
+		return nil, fmt.Errorf("doing rawConn Control: %w", err)
+	}
+
+	if err := authenticator.Authenticate(ucred); err != nil {
+		return nil, err
+	}
+
+	return ucred, nil
+}
+
+// SocketAuthenticator provides abstraction over various authentication types.
+type SocketAuthenticator interface {
+	Authenticate(ucred *unix.Ucred) error
+}
+
+type DaemonAuthenticator struct{}
+
+func NewDaemonAuthenticator() DaemonAuthenticator {
+	return DaemonAuthenticator{}
+}
+
+func (DaemonAuthenticator) Authenticate(ucred *unix.Ucred) error {
+	// root?
+	if ucred.Uid == 0 {
+		return nil
+	}
+
+	isGroup, err := isInAllowedGroup(ucred)
+	if err != nil {
+		return err
+	}
+
+	if !isGroup {
+		return ErrNoPermission
+	}
+
+	return nil
+}
+
+type FileshareAuthenticator struct {
+	controllingUserUUID uint32
+}
+
+func NewFileshareAuthenticator(controlingUserUUID uint32) FileshareAuthenticator {
+	return FileshareAuthenticator{
+		controllingUserUUID: controlingUserUUID,
+	}
+}
+
+func (f FileshareAuthenticator) Authenticate(ucred *unix.Ucred) error {
+	if ucred.Uid == 0 {
+		return nil
+	}
+
+	isGroup, err := isInAllowedGroup(ucred)
+	if err != nil {
+		return err
+	}
+
+	if !isGroup {
+		return ErrNoPermission
+	}
+
+	if ucred.Uid != f.controllingUserUUID {
+		return ErrNoPermission
+	}
+
+	return nil
+}
```

Comment:
Something like this can be done for now.
```suggestion
func (f FileshareAuthenticator) Authenticate(ucred *unix.Ucred) error {
        if err := DaemonAuthenticator{}.Authenticate(ucred); err != nil {
                return err
        }
        
	if ucred.Uid != f.controllingUserUUID {
		return ErrNoPermission
	}

	return nil
}
```
Or even full abstract authenticator wrapper